src/server/job.rs

```diff
@@ -206,45 +209,58 @@ impl Job {
         }
     }*/
 
-    pub fn set_running_state(&mut self, tako_task_id: TakoTaskId) {
+    pub fn set_running_state(&mut self, tako_task_id: TakoTaskId, worker_id: WorkerId) {
         let (_, state) = self.get_task_state_mut(tako_task_id);
         if matches!(state, JobTaskState::Waiting) {
-            *state = JobTaskState::Running;
+            *state = JobTaskState::Running(worker_id);
             self.counters.n_running_tasks += 1;
         }
     }
 
     pub fn set_finished_state(&mut self, tako_task_id: TakoTaskId) {
         let (_, state) = self.get_task_state_mut(tako_task_id);
-        assert!(matches!(state, JobTaskState::Running));
-        *state = JobTaskState::Finished;
-        self.counters.n_running_tasks -= 1;
-        self.counters.n_finished_tasks += 1;
+
+        match state {
+            JobTaskState::Running(worker_id) => {
+                *state = JobTaskState::Finished(*worker_id);
+                self.counters.n_running_tasks -= 1;
+                self.counters.n_finished_tasks += 1;
+            }
+            _ => assert!(matches!(state, JobTaskState::Running(_))),
```

Comment:
`panic!("Invalid worker state, expected Running, got {:?}", state)`

The same in set_failed_state.